### PR TITLE
[MSG] Made a RANGE_ADD mutation for FlatList

### PR DESCRIPTION
**Pull request by sarahscott**
This sends a new message and updates the UI accordingly: greyed out while sending and disabled text input field, and black once sent and re-enabled text input field.

Skip New Tests.

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  New dependencies added: react-native-reversed-flat-list.
  </td>
    </tr>
  </tbody>
</table>



<h2><a href="https://github.com/jevakallio/react-native-reversed-flat-list#readme">react-native-reversed-flat-list</a></h2>
<p>Author: Jani Eväkallio</p>
<p>Description: A reversed React Native FlatList, useful for chats and whatnot</p>
<p>Homepage: <a href="https://github.com/jevakallio/react-native-reversed-flat-list#readme">https://github.com/jevakallio/react-native-reversed-flat-list#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>9 days ago</td></tr><tr><td>Last Updated</td><td>9 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>2</td></tr><tr><td>Keywords</td><td>react, react-component, react-native, ios, android, chat, flatlist and listview</td></tr>
</table>

<details>
<summary><code>README</code></summary>
<h1 align="center">react-native-reversed-flat-list</h1>

<h4 align="center">
  A reversed React Native FlatList, useful for chats and whatnot
</h4>

***

## Documentation

Use exactly like you would use [FlatList](https://facebook.github.io/react-native/docs/flatlist.html). Docs to be written.

## Attribution

The basic mechanism is the same as [expo/react-native-invertible-scroll-view](https://github.com/expo/react-native-invertible-scroll-view), updated for FlatList.

Initial FlatList implementation was prototyped by [@joshyhargreaves](https://github.com/joshyhargreaves).

[MIT licensed](LICENSE)

</details>




  <details>
    <summary><code>yarn why react-native-reversed-flat-list</code> output</summary>
    <p><code><ul><li>Has been hoisted to "react-native-reversed-flat-list"</li><li>This module exists because it's specified in "dependencies".</li><li>Disk size without dependencies: "232kB"</li><li>Disk size with unique dependencies: "232kB"</li><li>Disk size with transitive dependencies: "232kB"</li><li>Number of shared dependencies: 0
    </li></ul></code></p>
  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by sarahscott** [comment]
This PR is on hold until @mzikherman and I's metaphysics/emission message-related updates are merged.

File: src/lib/Containers/Conversation.tsx [hunk 1]
```diff
@@ -110,19 +129,107 @@ export class Conversation extends React.Component<RelayProps, any> {
             <PlaceholderView />
           </HeaderTextContainer>
         </Header>
-        <MessagesList
-          data={messageData}
+        <ReversedFlatList
```
Comment (sarahscott) [inline src/lib/Containers/Conversation.tsx]: Going to remove this dependency; was an experiment.

**Comment by l2succes** [comment]
Note to self, update reaction-force once this is merged.
https://github.com/artsy/reaction-force/issues/172

**Comment by mzikherman** [comment]
bp

**Comment by alloy** [comment]
@mzikherman Dope!

Is it just the GIF (downscaling from retina) or is the top border of the input field actually missing once selected/entered text?

**Comment by mzikherman** [comment]
It's just the GIF artifact, it looks fine.

File: src/lib/Containers/Conversation.tsx [hunk 2]
```diff
@@ -87,12 +78,8 @@ export class Conversation extends React.Component<RelayProps, State> {
     // End of the list reached. Increase page size. Relay
     // will fetch only the required data to fill the new
     // page size.
-    if (item.key === this.state.messages.length - 1) {
-      this.props.relay.setVariables({ pageSize: this.state.messages.length + PAGE_SIZE }, readyState => {
-        if (readyState.done) {
-          debugger
-        }
-      })
+    if (item.key === this.props.me.conversation.messages.edges.length - 1) {
```
Comment (alloy) [inline src/lib/Containers/Conversation.tsx]: I get that this was sufficient as a a temp workaround, but it should be replaced with a proper scroll handler. Or possibly by using a pre-built one such as https://github.com/expo/react-native-invertible-scroll-view.

Comment (mzikherman) [inline src/lib/Containers/Conversation.tsx]: Yea

This was just something to get the subsequent Relay call for more data (+ re-render), and make sure it's working.

I'll look into (or would love to pair) on the more proper scroll behavior. I'm also a bit confused with this 'reversed flat list' thingie so maybe I can pair with @sarahscott 

Either way this worked!

File: src/lib/Components/Inbox/Conversations/Message.tsx [hunk 3]
```diff
@@ -79,14 +80,13 @@ export class Message extends React.Component<Props, any> {
...
     const date = moment(this.props.message.created_at).fromNow(true)
-
     return (
       <Container>
         <Avatar />
         <TextContainer>
           <Header>
             <SenderName>
-              {this.props.senderName}
+              {this.props.message.is_from_user ? this.props.userName : this.props.partnerName}
```
Comment (mzikherman) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: This just felt like more of a `Message` concern (rendering the right label based on the best guess as to who sent that particular msg).

File: src/lib/Containers/Conversation.tsx [hunk 4]
```diff
@@ -102,21 +115,96 @@ export class Conversation extends React.Component<RelayProps, any> {
...
           data={messages}
           renderItem={this.renderMessage.bind(this)}
+          length={messages.length}
           ItemSeparatorComponent={DottedBorder}
         />
         <ComposerContainer>
-          <Composer />
+          <Composer
+            onSubmit={text => {
+              this.props.relay.commitUpdate(
```
Comment (mzikherman) [inline src/lib/Containers/Conversation.tsx]: This works~!

File: src/lib/relay/config.tsx [hunk 5]
```diff
@@ -13,6 +13,8 @@ if (Emission && Emission.useStagingEnvironment) {
   gravityURL = "https://api.artsy.net"
 }
 
+metaphysicsURL = "http://localhost:5001"
```
Comment (mzikherman) [inline src/lib/relay/config.tsx]: 🌴

File: package.json [hunk 6]
```diff
@@ -29,30 +28,21 @@
     "tslint": "tslint",
     "sync-colors": "cd externals/elan && git fetch && git checkout origin/master && cp components/lib/variables/colors.json ../../data",
     "sync-externals": "npm run-script sync-schema && npm run-script sync-colors",
-    "sync-schema": "cd externals/metaphysics && git fetch && git checkout origin/master && yarn install && npm run dump-schema -- ../../data && rm -rf node_modules",
+    "sync-schema": "cd externals/metaphysics && git fetch && git checkout origin/sarah-improved-message-mutation && yarn install && npm run dump-schema -- ../../data && rm -rf node_modules",
```
Comment (l2succes) [inline package.json]: Note to put `master` back after the MP branch is merged

Comment (alloy) [inline package.json]: Yup 👍

**Comment by sarahscott** [comment]
🔥

File: src/lib/relay/config.tsx (hunk 5 above)
Comment (alloy) [inline src/lib/relay/config.tsx]: Gah!! This is still in

File: src/lib/Containers/Conversation.tsx [hunk 7]
```diff
@@ -61,7 +63,24 @@ const ComposerContainer = styled.View`
   marginLeft: 20
 `
 
-export class Conversation extends React.Component<RelayProps, any> {
+const PAGE_SIZE = 100
```
Comment (mzikherman) [inline src/lib/Containers/Conversation.tsx]: 🌞 

So proper scroll/pagination post-beta yes?

Comment (alloy) [inline src/lib/Containers/Conversation.tsx]: Aye :+1:

Comment (alloy) [inline src/lib/Containers/Conversation.tsx]: https://github.com/artsy/emission/issues/642

**Pull request merged**